galley/pkg/crd/synchronizer.go

```diff
@@ -0,0 +1,364 @@
+		}
+
+		s.queue.Done(item)
+		if s.eventHook != nil {
+			s.eventHook(item)
+		}
+		log.Debugf("marked item as done: %v", item)
+	}
+}
+
+// processQueueItem is a panic-protected processor of work items.
+func (s *Synchronizer) processQueueItem(item interface{}) (result bool) {
+	reachedEnd := false
+	defer func() {
+		r := recover()
+		if !reachedEnd {
+			log.Errorf("panic in crd.Synchronizer.processChange: %v", r)
+			// TODO: This will cause the panic to keep recurring in a rate-limited way. This will cause many log
+			// entries, which would be helpful during initial dev cycle. Eventually we should  flip this to true,
+			// as we should not try to retry work that is causing the panic.
+			result = false
+		}
+	}()
+
+	if info, ok := item.(*change.Info); !ok {
+		log.Errorf("Got a non-change item from the queue: %v", item)
+		result = true
+	} else {
+		result = s.processChange(info)
+	}
+
+	reachedEnd = true
+	return
+}
+
+// processChange does the actual processing of a change notification. Returns true, if the rate-limiter should
+// forget the work and not reschedule again.
+func (s *Synchronizer) processChange(info *change.Info) (forget bool) {
+
+	sourceGv, destinationGv, found := s.mapping.GetGroupVersion(GroupOf(info.Name))
+	if !found {
+		log.Debugf("skipping unrelated CRD change notification: '%v'", info.Name)
+		forget = true
+		return
+	}
+
+	sourceName := rewriteName(destinationGv.Group, info.Name)
+	destinationName := rewriteName(sourceGv.Group, info.Name)
+
+	var err error
+	var sourceCrd *apiext.CustomResourceDefinition
+	var destinationCrd *apiext.CustomResourceDefinition
+
+	if sourceCrd, err = s.getCrd(destinationName); err != nil {
+		return
+	}
+
+	if destinationCrd, err = s.getCrd(sourceName); err != nil {
+		return
+	}
+
+	forget = true
+	if sourceCrd != nil {
+		candidate := rewrite(sourceCrd, destinationGv.Group, destinationGv.Version)
+
+		// Update the destination CRD, if it is different than the source one.
+		if destinationCrd != nil {
+			candidate.ResourceVersion = destinationCrd.ResourceVersion
+
+			if !equals(candidate, destinationCrd) {
+				log.Infof("Updating destination CRD: %s => %s (versions: %s => %s)",
+					sourceCrd.Name, candidate.Name, sourceCrd.ResourceVersion, candidate.ResourceVersion)
+				if _, err = s.crdi.Update(candidate); err != nil {
+					log.Errorf("Error during CRD update: name='%s', err:'%v'", sourceName, err)
+					forget = false
+				}
+			}
+		} else {
+			// Insert a new CRD in destination, as it doesn't exist.
+			log.Infof("Creating destination CRD: %s => %s (source version: %s)",
+				sourceCrd.Name, candidate.Name, sourceCrd.ResourceVersion)
+			if _, err = s.crdi.Create(candidate); err != nil {
+				// TODO: Consider doing an update here if it is found to exists.
+				log.Errorf("Error during CRD create: name='%s', err:'%v'", sourceName, err)
+				forget = false
+			}
+		}
+	} else {
+		// The source CRD doesn't exist, we need to take corrective action
```

Comment:
Did you look into whether garbage collection would serve the same purpose? See https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection.

Reply:
I did. I didn't want to hijack this for Galley purposes (I've been trying to make the "semantical foot-print" of the synchronization op small). It maybe a good feature to depend on once we make this distribution mechanism fully internal..